### DISCUSSION TITLE
Sampling frequency timers

### DISCUSSION THREAD

**Discussion by martareyp** [TheAlbertDev/MASB#61] [Q&A] [answered]
Hi Albert,
we are working on timers and we have found two functions __HAL_TIM_SET_PRESCALER and __HAL_TIM_SET_COUNTER. We would like to use them to introduce the sampling period we read from the CA_configuration but we are not sure where we should introduce it. We have look at the formulas of practice 3 and at the documentation but we are not able to know how to obtain a prescaler value and counter value from the sampling period. Could you advise us with which procedure or formula follow? Are these functions a good option for setting the sampling period?
Thank you in advance,
Marta and Helena

**Comment by Albert-Alvarez** [comment, marked as answer]
We discussed it in a virtual meeting, but I put it here for everyone.

To change the timer period in the code, there is the function `__HAL_TIM_SET_AUTORELOAD()`. You have information about it in the [HAL documentation](https://www.st.com/resource/en/user_manual/dm00105879-description-of-stm32f4-hal-and-ll-drivers-stmicroelectronics.pdf) (page 1064).

With this function we can change the timer period. To avoid having to modify the prescaler, whose function I already described in another issue (#27), take in the `.ioc` file a preescaler that gives you a clock frequency with a period that is divisor of 1 ms. In this way, just touching the period will be enough. For example, with a prescaler of 8,399, you get a clock frequency of 10 kHz. In other words, each increment of the timer is 0.1 ms. Since the timer counter is 16 bits (in the case of TIM3), you can count from 0 to 65,535 or, in other words, from 0 to 6.5535 seconds with a resolution of 0.1 ms.
 
To change the period in the code would be:

```C
// Se supone que detenemos el timer justo al finalizar la medicion anterior. Por
// eso esta comentada.
// HAL_TIM_Base_Stop_IT(&htim3); 

__HAL_TIM_SET_AUTORELOAD(&htim3, samplingPeriodMs * 10); // Fijamos el periodo.
// El mutliplicar el samplingPeriodMs por 10 para fijar el periodo solo es valido
// si se fija una frecuencia de trabajo para el timer de 10 kHz.

__HAL_TIM_SET_COUNTER(&htim3, 0); // Reiniciamos el contador del timer a 0.
HAL_TIM_Base_Start_IT(&htim3); // Iniciamos el timer.
```

I do not put the `HAL_TIM_Base_Stop_IT()` function since I assume that you call it as soon as you finish a measurement.
